docs/proposals/client-package-structure.md

```diff
@@ -0,0 +1,153 @@
+# Client: layering and package structure
+
+## Desired layers
+
+### Transport
+
+The transport layer is concerned with round-tripping requests to an apiserver
+somewhere. It consumes a Config object with options appropriate for this.
+(That's most of the current client.Config structure.)
+
+Third parties (e.g., open shift) may have their own implementation.
+
+Transport delivers an object that implements http's RoundTripper interface
+and/or can be used in place of http.DefaultClient to route requests.
+
+Transport objects are safe for concurrent use, and are cached and reused by
+subsequent layers.
+
+Tentative name: "Transport".
+
+Action items:
+* Split out of current client package into a new package. (@krousey)
+
+### RESTClient/request.go
+
+RESTClient consumes a Transport, a group, and a version, and produces something
+that implements the interface currently in request.go. That is, with a
+RESTClient, you can write chains of calls like: c.Get().Path(p).Param("name",
+"value").Do()
+
+The group and version may be empty when constructing a RESTClient. This is valid
+for executing discovery commands. The group and version may be overridable with
+a function call.
+
+RESTClients may have semantic information built in. For example, if the
+query parameter changes from "labels" to "label-selector", when you call the
+RESTClient construction logic, it should give you a RESTClient which uses the
+appropriate query parameter when you call the "LabelSelector()" method.
+
+3rd parties such as open shift may provide their own implementation, but it
+should conform to the same interface as the standard RESTClient.
+
+Action items:
+* Split out of the current client package. (@krousey)
+* Convert to an interface (currently, it's a struct).
+
+### Mux layer
+
+The client muxer layer has a map of group/version to cached RESTClient, and
+knows how to construct a new RESTClient in case of a cache miss.
+
+Tentative name: MetaClient
+
+Action items:
+* Move client cache out of kubectl libraries into a more general home.
```

Comment:
Well, I called the thing that talks to the group/version/resource page the "discovery client". I picture this thing using a discovery client to populate its cache on an as-needed basis.